clients/src/main/java/org/apache/kafka/common/security/ssl/DefaultSslEngineFactory.java

```diff
@@ -281,7 +295,7 @@ public boolean shouldBeRebuilt(Map<String, Object> nextConfigs) {
          * @throws KafkaException if the file could not be read or if the keystore could not be loaded
          *   using the specified configs (e.g. if the password or keystore type is invalid)
          */
-        KeyStore load() {
+        private KeyStore load() {
             try (InputStream in = Files.newInputStream(Paths.get(path))) {
```

Comment:
Please make it possible to load the keystore not from a file but e.g. from a variable. This could be easily done by providing a protected method "byte[] getKeyStoreContent" (or similar). The default impl. would read the contents from the file. But I caould subclass the default impl. and just use a property set from outside.

The same applies for the truststore.

Reply:
@fml2 Currently the DefaultSslEngineFactory class is 'final' so you won't be able to override. I wrote my custom implementation for our needs for SslEngineFactory, primary reason for the KIP, and it resulted in total of 246 LOC with imports and 213 LOC without imports. I think it is reasonably good to have such a great flexibility with just 200+ LOC. Can you please try writing your custom impl (may be just by copying DefaultSslEngineFactory to start with) and see if it serves your needs? I completely agree that there will be duplicate code for your custom implementation but unless we get enough convincing case on making some methods overridable it is better to keep it the way it is.

Reply:
@maulin-vasavada Do you know the reason for making the class final? What is the rationale? Why make the developer's life harder than it is? :-)

Reply:
@fml2 Your requirements are byte[], somebody else might require PEM format string. Sometimes you could have List<X509..cert> objects for Truststore etc. Due to various formats it is not easy to generalize it in a way which will fit every possible use-case if we expose the extension the way you are suggesting. Ultimately I agree it is all byte[] but the formats matter.

Reply:
DefaultSslEngineFactory is an internal class. We don't encourage extending this for custom implementations since there is no guarantee that we won't change it in future. For custom implementations of SslEngineFactory, it would be better to rely on the interface.

Reply:
OK, I see. But it's a pity I have to write/copy more than 200 lines of code to just have a functionality many systems (I'm quite sure) need.